File: smart-toast/src/main/java/com/coder/zzq/smartshow/toast/classic/ClassicToastInvoker.java
package com.coder.zzq.smartshow.toast.classic;

import android.view.Gravity;
import android.widget.Toast;

import com.coder.zzq.smartshow.toast.Constants;
import com.coder.zzq.smartshow.toast.PlainToastApi;
import com.coder.zzq.smartshow.toast.compact.CompactToast;
import com.coder.zzq.smartshow.toast.schedule.ToastScheduler;
import com.coder.zzq.toolkit.Utils;

public class ClassicToastInvoker implements ClassicToastView.Overall, ClassicToastView.ConfigSetter {
    private final ClassicToast.Config mConfig = new ClassicToast.Config();

    @Override
    public ClassicToastView.ConfigSetter config() {
        return this;
    }

    @Override
    public void show(CharSequence msg) {
        showHelper(msg, Gravity.BOTTOM | Gravity.CENTER_HORIZONTAL, 0, Constants.DEFAULT_VALUE, Toast.LENGTH_SHORT);
    }

    @Override
    public void show(int msg) {
        show(Utils.getStringFromRes(msg));
    }

    @Override
    public void showAtTop(CharSequence msg) {
        showHelper(msg, Gravity.TOP | Gravity.CENTER_HORIZONTAL, 0, Utils.getToolbarHeight() + Utils.dpToPx(40), Toast.LENGTH_SHORT);
    }

    @Override
    public void showAtTop(int msg) {
        showAtTop(Utils.getStringFromRes(msg));
    }

    @Override
    public void showInCenter(CharSequence msg) {
        showHelper(msg, Gravity.CENTER, 0, 0, Toast.LENGTH_SHORT);
    }

    @Override
    public void showInCenter(int msg) {
        showInCenter(Utils.getStringFromRes(msg));
    }

    @Override
    public void showAtLocation(CharSequence msg, int gravity, float xOffsetDp, float yOffsetDp) {
        showHelper(msg, gravity, Utils.dpToPx(xOffsetDp), Utils.dpToPx(yOffsetDp), Toast.LENGTH_SHORT);
    }

    @Override
    public void showAtLocation(int msg, int gravity, float xOffsetDp, float yOffsetDp) {
        showAtLocation(Utils.getStringFromRes(msg), gravity, xOffsetDp, yOffsetDp);
    }

    @Override
    public void showLong(CharSequence msg) {
        showHelper(msg, Gravity.BOTTOM | Gravity.CENTER_HORIZONTAL, 0, Constants.DEFAULT_VALUE, Toast.LENGTH_LONG);
    }

    @Override
    public void showLong(int msg) {
        showLong(Utils.getStringFromRes(msg));
    }

    @Override
    public void showLongAtTop(CharSequence msg) {
        showHelper(msg, Gravity.TOP | Gravity.CENTER_HORIZONTAL, 0, Utils.getToolbarHeight() + Utils.dpToPx(40), Toast.LENGTH_LONG);
    }

    @Override
    public void showLongAtTop(int msg) {
        showLongAtTop(Utils.getStringFromRes(msg));
    }

    @Override
    public void showLongInCenter(CharSequence msg) {
        showHelper(msg, Gravity.CENTER, 0, 0, Toast.LENGTH_LONG);
    }

    @Override
    public void showLongInCenter(int msg) {
        showLongInCenter(Utils.getStringFromRes(msg));
    }

    @Override
    public void showLongAtLocation(CharSequence msg, int gravity, float xOffsetDp, float yOffsetDp) {
        showHelper(msg, gravity, Utils.dpToPx(xOffsetDp), Utils.dpToPx(yOffsetDp), Toast.LENGTH_LONG);
    }

    @Override
    public void showLongAtLocation(int msg, int gravity, float xOffsetDp, float yOffsetDp) {
        showLongAtLocation(Utils.getStringFromRes(msg), gravity, xOffsetDp, yOffsetDp);
    }

    private void showHelper(CharSequence msg, int gravity, int xOffsetDp, int yOffsetDp, int duration) {
        mConfig.mMsg = msg;
        mConfig.mGravity = gravity;
        mConfig.mXOffset = xOffsetDp;
        mConfig.mYOffset = yOffsetDp;
        mConfig.mDuration = duration;
        ToastScheduler.get().schedule(new CompactToast(ClassicToastFactory.get(), mConfig));
    }

    @Override
    public ClassicToastView.ConfigSetter transition(boolean b) {
        mConfig.mTransition = b;
        return this;
    }

    @Override
    public ClassicToastView.ConfigSetter cancelOnActivityExit(boolean b) {
        mConfig.mCancelOnActivityExit = b;
        return this;
    }

    @Override
    public ClassicToastView.ConfigSetter backgroundColor(int color) {
        mConfig.mBackgroundColor = color;
        return this;
    }

    @Override
    public ClassicToastView.ConfigSetter backgroundColorResource(int color) {
        backgroundColor(Utils.getColorFromRes(color));
        return this;
    }

    @Override
    public ClassicToastView.ConfigSetter backgroundDrawableResource(int background) {
        mConfig.mBackgroundResource = background;
        return this;
    }

    @Override
    public ClassicToastView.ConfigSetter msgColor(int color) {
        mConfig.mMsgColor = color;
        return this;
    }

    @Override
    public ClassicToastView.ConfigSetter msgColorResource(int color) {
        msgColor(Utils.getColorFromRes(color));
        return this;
    }

    @Override
    public ClassicToastView.ConfigSetter msgSize(float size) {
        mConfig.mMsgSize = size;
        return this;
    }

    @Override
    public ClassicToastView.ConfigSetter msgBold(boolean b) {
        mConfig.mMsgBold = b;
        return this;
    }

    @Override
    public ClassicToastView.ConfigSetter iconResource(int icon) {
        mConfig.mIcon = icon;
        return this;
    }

    @Override
    public ClassicToastView.ConfigSetter iconPosition(int position) {
        mConfig.mIconPosition = position;
        return this;
    }

    @Override
    public ClassicToastView.ConfigSetter iconSizeDp(float iconSizeDp) {
        mConfig.mIconSizeDp = Utils.dpToPx(iconSizeDp);
        return this;
    }

    @Override
    public ClassicToastView.ConfigSetter iconPaddingDp(float iconPaddingDp) {
        mConfig.mIconPaddingDp = Utils.dpToPx(iconPaddingDp);
        return this;
    }

    @Override
    public PlainToastApi apply() {
        return this;
    }
}

File: smart-toast/src/main/java/com/coder/vincent/smart_toast/alias/classic/ClassicToastInvoker.kt
package com.coder.vincent.smart_toast.alias.classic

import android.graphics.drawable.Drawable
import android.view.Gravity
import androidx.annotation.ColorInt
import androidx.annotation.ColorRes
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import com.coder.vincent.series.common_lib.Toolkit
import com.coder.vincent.series.common_lib.bean.IconPosition
import com.coder.vincent.series.common_lib.bean.TextStyle
import com.coder.vincent.series.common_lib.dpToPx
import com.coder.vincent.series.common_lib.resourceToColor
import com.coder.vincent.series.common_lib.resourceToDrawable
import com.coder.vincent.series.common_lib.resourceToString
import com.coder.vincent.smart_toast.DEFAULT_TOAST_Y_OFFSET
import com.coder.vincent.smart_toast.ShowToast
import com.coder.vincent.smart_toast.TOAST_ALIAS_CLASSIC
import com.coder.vincent.smart_toast.bean.Duration
import com.coder.vincent.smart_toast.factory.Location
import com.coder.vincent.smart_toast.schedule.ToastScheduler

class ClassicToastInvoker : ClassicToastFacade.Overall, ClassicToastFacade.ConfigSetter {
    private val config: ClassicToast.Config =
        ClassicToast.Config().apply { alias = TOAST_ALIAS_CLASSIC }

    override fun config(): ClassicToastFacade.ConfigSetter = this

    override fun show(msg: CharSequence) {
        showHelper(
            msg, config.duration, Location(
                Gravity.BOTTOM or
                        Gravity.CENTER_HORIZONTAL, 0, DEFAULT_TOAST_Y_OFFSET
            )
        )
    }

    override fun show(@StringRes msg: Int) {
        show(msg.resourceToString())
    }

    override fun showAtTop(msg: CharSequence) {
        showHelper(
            msg, config.duration,
            Location(
                Gravity.TOP or
                        Gravity.CENTER_HORIZONTAL, 0, Toolkit.getToolbarHeight() + 40f.dpToPx()
            ),
        )
    }

    override fun showAtTop(@StringRes msg: Int) {
        showAtTop(msg.resourceToString())
    }

    override fun showInCenter(msg: CharSequence) {
        showHelper(msg, config.duration, Location(Gravity.CENTER, 0, 0))
    }

    override fun showInCenter(@StringRes msg: Int) {
        showInCenter(msg.resourceToString())
    }

    override fun showAtLocation(
        msg: CharSequence,
        gravity: Int,
        xOffsetDp: Float,
        yOffsetDp: Float,
    ) {
        showHelper(
            msg,
            config.duration,
            Location(gravity, xOffsetDp.dpToPx(), yOffsetDp.dpToPx())
        )
    }

    override fun showAtLocation(
        @StringRes msg: Int,
        gravity: Int,
        xOffsetDp: Float,
        yOffsetDp: Float,
    ) {
        showAtLocation(msg.resourceToString(), gravity, xOffsetDp, yOffsetDp)
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun showLong(msg: CharSequence) {
        showHelper(
            msg, Duration.LONG,
            Location(
                Gravity.BOTTOM or
                        Gravity.CENTER_HORIZONTAL, 0, DEFAULT_TOAST_Y_OFFSET
            ),
        )
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun showLong(@StringRes msg: Int) {
        showLong(msg.resourceToString())
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun showLongAtTop(msg: CharSequence) {
        showHelper(
            msg, Duration.LONG, Location(
                Gravity.TOP or Gravity.CENTER_HORIZONTAL, 0,
                Toolkit.getToolbarHeight() + 40f.dpToPx()
            )
        )
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun showLongAtTop(@StringRes msg: Int) {
        showLongAtTop(msg.resourceToString())
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun showLongInCenter(msg: CharSequence) {
        showHelper(msg, Duration.LONG, Location(Gravity.CENTER, 0, 0))
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun showLongInCenter(@StringRes msg: Int) {
        showLongInCenter(msg.resourceToString())
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun showLongAtLocation(
        msg: CharSequence,
        gravity: Int,
        xOffsetDp: Float,
        yOffsetDp: Float,
    ) {
        showHelper(
            msg,
            Duration.LONG,
            Location(gravity, xOffsetDp.dpToPx(), yOffsetDp.dpToPx())
        )
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun showLongAtLocation(
        @StringRes msg: Int,
        gravity: Int,
        xOffsetDp: Float,
        yOffsetDp: Float,
    ) {
        showLongAtLocation(msg.resourceToString(), gravity, xOffsetDp, yOffsetDp)
    }

    private fun showHelper(
        msg: CharSequence,
        duration: Duration,
        location: Location,
    ) {
        config.message = msg
        config.location = location
        config.duration = duration
        ToastScheduler.schedule(config, ClassicToastFactory)
    }

    override fun backgroundDrawable(bgDrawable: Drawable): ClassicToastFacade.ConfigSetter =
        this.apply {
            config.backgroundDrawable = bgDrawable
        }

    override fun backgroundResource(@DrawableRes bgResource: Int): ClassicToastFacade.ConfigSetter =
        this.apply {
            config.backgroundDrawable = bgResource.resourceToDrawable()!!
        }

    override fun messageStyle(
        @ColorInt color: Int,
        size: Float,
        bold: Boolean,
    ): ClassicToastFacade.ConfigSetter = this.apply {
        config.messageStyle = TextStyle(color, size, bold)
    }

    override fun iconDrawable(iconDrawable: Drawable?): ClassicToastFacade.ConfigSetter =
        this.apply {
            config.iconDrawable = iconDrawable
        }

    override fun iconResource(@DrawableRes iconResource: Int): ClassicToastFacade.ConfigSetter =
        this.apply {
            config.iconDrawable = iconResource.resourceToDrawable()
        }

    override fun iconSize(size: Float?): ClassicToastFacade.ConfigSetter = this.apply {
        config.iconSize = size
    }

    override fun marginBetweenIconAndMsg(margin: Float): ClassicToastFacade.ConfigSetter =
        this.apply {
            config.marginBetweenIconAndMsg = margin
        }

    override fun backgroundColor(backgroundColor: Int): ClassicToastFacade.ConfigSetter =
        this.apply {
            config.backgroundColor = backgroundColor
        }

    override fun backgroundColorResource(@ColorRes backgroundColorResource: Int):
            ClassicToastFacade.ConfigSetter = this.apply {
        config.backgroundColor = backgroundColorResource.resourceToColor()
    }

    override fun iconPosition(iconPosition: IconPosition): ClassicToastFacade.ConfigSetter =
        this.apply {
            config.iconPosition = iconPosition
        }

    override fun duration(duration: Duration): ClassicToastFacade.ConfigSetter =
        this.apply {
            config.duration = duration
        }


    override fun commit(): ShowToast = this
}
